docs/proposed/http-introducer-internals.rst

```diff
@@ -0,0 +1,192 @@
+This would sensibly reflect the fact that two properties remain the same:
+
+* The URI has the same self-certifying capability nature as the Foolscap URL is supersedes.
+  "tubid" has been replaced with the SPKI hash which serves a parallel purpose.
+  "swissnum" remains an unguessable string which imues the URI with the capability nature.
+* It refers to the same underlying service.
+  Only the protocol is changing.
+
+It is common to think of a URI scheme as identifying a protocol
+(https, wss, xmpp, etc).
+However this only reflects the fact that the vast majority of resources are accessible via only one protocol and so that protocol and the resource itself become conflated.
+There *are* URI schemes which are protocol agnostic
+(mailto, im, pres, etc).
+The "https" example is also something of a trick.
+The resource identified by an "https"-scheme URI may be reachable via any one (or more!) of several different HTTP-family protocols
+(0.9, 1.0, 1.1, 2.0, and likely soon 3.0).
+
+In implementation terms the complication introduced by using "pb" is that the client is not given a signal as to which protocol to use to attempt to interact with the resource.
+This could be addressed in one of at least two ways:
+
+1. Try all protocols the client supports concurrently and allow the attempts with the incorrect protocol to fail gracefully.
+   The result of this could also be remembered to avoid the need for such concurrent efforts on all future connection attempts.
+2. Use one of the protocol negotiation features of TLS (eg ALPN).
+   This is exactly the mechanism used to negotiate the version of the HTTP protocol.
+
+pb+http
+```````
+
+HTTP Introducer could continue to use *pb* but with a hint about its divergence from the default/historical wire protocol.
+This is similar to the "git+https" scheme.
+An advantage of this scheme is that it makes the protocol explicit and removes the need for further negotiation.
+A disadvantage is that since it forces the new HTTP protocol it requires two different URIs if there is to be a period where both protocols are offered.
+The two URIs must each be transmitted to to clients somehow.
+Clients must persist both of them.
+etc.
+
+This also lends itself well to future protocol transitions.
+For example,
+we might have *pb+wss* or *pb+X* for some yet-to-be-invented *X*.
+
+
+x-ocap
+``````
+
+Or some other nice-looking brand new string.
+
+This option has the least obvious semantics.
+In inventing an entirely new scheme,
+we have to invent all of its behavior.
+
+This could be like the *pb* option but with a new name and no baggage from Foolscap.
+As soon as we decide to make another protocol change then we're back in the same place as we are now.
+The only difference is that the "HTTP Introducer" protocol would play the part Foolscap now plays.
+
+For the purposes of migrating from Foolscap it has the same drawback as *pb+http*:
+there are two URIs clients must be made aware of.
```

Comment:
New clients need the new-style URI and old clients need the old-style URI -- is that what you mean here?

Reply:
Oh, also I got to here and re-remembered that this has "internals" versus "high-level" document; maybe each should point out the existence of the other.